torch/_higher_order_ops/templated_attention.py

```diff
@@ -220,3 +249,383 @@ def templated_attention_fake_tensor_mode(
+        fw_args = ctx.saved_tensors
+        query, key, value, out, logsumexp, *other_buffers = fw_args
+        fw_graph = ctx._fw_graph
+        joint_graph = ctx._joint_graph
+        # We have asserted that other_buffers do not require grad in the forward
+        none_grads = [None] * (2 + len(other_buffers))
+        grad_query, grad_key, grad_value = templated_attention_backward(
+            query,
+            key,
+            value,
+            out,
+            logsumexp,
+            grad_out,
+            fw_graph,
+            joint_graph,
+            *other_buffers,
+        )
+        return grad_query, grad_key, grad_value, *none_grads
+
+
+@templated_attention.py_impl(DispatchKey.Autograd)
+def templated_attention_autograd(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: Tuple[torch.Tensor, ...],
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    input_requires_grad = any(t.requires_grad for t in (query, key, value))
+    if torch.is_grad_enabled() and input_requires_grad:
+        example_vals = [
+            torch.zeros((), dtype=query.dtype, requires_grad=input_requires_grad)
+        ] + [torch.zeros((), dtype=torch.int) for _ in range(4)]
+        fw_graph, bw_graph = create_fw_bw_graph(score_mod, example_vals, other_buffers)
+    else:
+        fw_graph, bw_graph = score_mod, None
+    out, logsumexp = TemplatedAttentionAutogradOp.apply(
+        query, key, value, fw_graph, bw_graph, *other_buffers
+    )
+    return out, logsumexp
+
+
+# ---------------------------- Backward HOP Implementation ----------------------------
+
+
+@templated_attention_backward.py_impl(DispatchKey.CompositeExplicitAutograd)
+def sdpa_dense_backward(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    logsumexp: torch.Tensor,
+    grad_out: torch.Tensor,
+    fw_graph: Callable,  # GraphModule type hint?
+    joint_graph: Callable,
+    *other_buffers: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    working_precision = torch.float64 if query.dtype == torch.float64 else torch.float32
+    scores = (query @ key.transpose(-2, -1)).to(working_precision)
+
+    b = torch.arange(0, scores.size(0), device=scores.device)
+    h = torch.arange(0, scores.size(1), device=scores.device)
+    m = torch.arange(0, scores.size(2), device=scores.device)
+    n = torch.arange(0, scores.size(3), device=scores.device)
+
+    in_dim_buffers = (None,) * len(other_buffers)
+    score_mod = torch.vmap(fw_graph, in_dims=(0, None, None, None, 0) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, None, 0, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, 0, None, None, None) + in_dim_buffers)
+
+    scores = score_mod(scores, b, h, m, n, *other_buffers).to(working_precision)
```

Comment:
Change it to like... `scores_mod = score_mod(scores)`

and change the subsequent usage on 413 to use `scores_mod`